rust/connlib/tunnel/src/peer/filter_engine.rs

```diff
@@ -64,13 +64,23 @@ impl AllowRules {
     }
 
     fn apply(&self, protocol: Result<Protocol, UnsupportedProtocol>) -> Result<(), Filtered> {
-        match protocol? {
-            Protocol::Tcp(port) if self.tcp.contains(&port) => Ok(()),
-            Protocol::Udp(port) if self.udp.contains(&port) => Ok(()),
-            Protocol::Icmp(_) if self.icmp => Ok(()),
-            Protocol::Tcp(_) => Err(Filtered::Tcp),
-            Protocol::Udp(_) => Err(Filtered::Udp),
-            Protocol::Icmp(_) => Err(Filtered::Icmp),
+        match protocol {
+            Ok(Protocol::Tcp(port)) if self.tcp.contains(&port) => Ok(()),
+            Ok(Protocol::Udp(port)) if self.udp.contains(&port) => Ok(()),
+            Ok(Protocol::Icmp(_)) if self.icmp => Ok(()),
+
+            // If ICMP is allowed, we don't care about the specific ICMP type.
+            // i.e. it doesn't have to be an echo request / reply.
+            Err(
+                UnsupportedProtocol::UnsupportedIcmpv4Type(_)
+                | UnsupportedProtocol::UnsupportedIcmpv6Type(_),
+            ) if self.icmp => Ok(()),
```

Comment:
This is a defense-in-depth kind of change: The other change ensures we don't actually call this filter at all for ICMP errors. Yet, I still opted to fix this here to ensure it is correct in isolation, i.e. IF we apply the filter, we allow any kind of ICMP message, not just echo requests and replies.

This is as correct as we can be in here. We don't know whether the entire packet is an ICMP error or not so from the perspective of this component, all we know is whether we should allow ICMP traffic.